File: Project32/backend/api/minio_service.py
import uuid
from minio import Minio
from django.conf import settings


class MinioService:
    def __init__(self):
        self.client = Minio(
            settings.MINIO_ENDPOINT,
            access_key=settings.MINIO_ACCESS_KEY,
            secret_key=settings.MINIO_SECRET_KEY,
            secure=settings.MINIO_SECURE
        )
        self.bucket = settings.MINIO_VIDEO_BUCKET
        self._ensure_bucket()

    def _ensure_bucket(self):
        try:
            if not self.client.bucket_exists(self.bucket):
                self.client.make_bucket(self.bucket)
        except Exception:
            pass

    def upload_file(self, file, file_name=None):
        if file_name is None:
            file_name = file.name
        
        object_name = f"{uuid.uuid4()}_{file_name}"
        
        self.client.put_object(
            self.bucket,
            object_name,
            file,
            length=file.size,
            content_type=file.content_type
        )
        
        return object_name

    def get_file_url(self, object_name, expires=3600):
        return self.client.presigned_get_object(self.bucket, object_name, expires=expires)

    def delete_file(self, object_name):
        self.client.remove_object(self.bucket, object_name)

    def list_files(self):
        return self.client.list_objects(self.bucket)

File: Project21/backend/src/modules/minio/minio.service.ts
import { Injectable, OnModuleInit } from '@nestjs/common';
import * as Minio from 'minio';

@Injectable()
export class MinioService implements OnModuleInit {
  private minioClient: Minio.Client;
  private readonly bucketName = process.env.MINIO_BUCKET || 'heritage-courseware';

  async onModuleInit() {
    this.minioClient = new Minio.Client({
      endPoint: process.env.MINIO_ENDPOINT || 'localhost',
      port: parseInt(process.env.MINIO_PORT || '9000'),
      useSSL: false,
      accessKey: process.env.MINIO_ACCESS_KEY || 'minioadmin',
      secretKey: process.env.MINIO_SECRET_KEY || 'minioadmin123',
    });

    await this.ensureBucketExists();
  }

  private async ensureBucketExists() {
    try {
      const exists = await this.minioClient.bucketExists(this.bucketName);
      if (!exists) {
        await this.minioClient.makeBucket(this.bucketName, 'us-east-1');
        const policy = {
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Principal: { AWS: ['*'] },
              Action: ['s3:GetObject'],
              Resource: [`arn:aws:s3:::${this.bucketName}/*`],
            },
          ],
        };
        await this.minioClient.setBucketPolicy(this.bucketName, JSON.stringify(policy));
      }
    } catch (e) {
      console.log('MinIO bucket init:', e.message);
    }
  }

  async uploadFile(objectName: string, file: Express.Multer.File): Promise<{ url: string; etag: string }> {
    const metaData = {
      'Content-Type': file.mimetype,
    };

    const result = await this.minioClient.putObject(
      this.bucketName,
      objectName,
      file.buffer,
      file.size,
      metaData,
    );

    const url = await this.getFileUrl(objectName);
    return { url, etag: result.etag };
  }

  async getFileUrl(objectName: string): Promise<string> {
    const endpoint = process.env.MINIO_ENDPOINT || 'localhost';
    const port = process.env.MINIO_PORT || '9000';
    return `http://${endpoint === 'minio' ? 'localhost' : endpoint}:${port}/${this.bucketName}/${objectName}`;
  }

  async deleteFile(objectName: string): Promise<void> {
    await this.minioClient.removeObject(this.bucketName, objectName);
  }

  getBucketName(): string {
    return this.bucketName;
  }
}
